src/main/java/io/bittiger/ads/activity/AdsDao.java

```diff
@@ -39,23 +41,77 @@ public void testMemcached() throws IOException {
         System.out.println(object);
     }
 
+
+    /****** Inverted Index ******/
+    public Set<Ad> getAds(String key) {
+        try {
+            return (Set<Ad>) getCache().get(key);
+        } catch (IOException e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    /****** Forward Index ******/
     public Ad getAd(long key) {
         try {
-            return (Ad)getCache().get(Long.toString(key));
+            Ad ad = (Ad)getCache().get("fwd" + Long.toString(key));
+            return ad;
         } catch (IOException e) {
             e.printStackTrace();
             return null;
         }
     }
 
+    public Set<Ad> traverseFwdIndex(String keyword) {
+        Set<Ad> ads = new HashSet<Ad>();
+            /* do nothing for now
+            since spymemcached does not support traversing all the keys,
+            this method will be used when we setup an independent
+            database/HashMap which supports traversing keys */
+        return ads;
+    }
+
     public boolean setAd(Ad ad) {
         String key = Long.toString(ad.getAdId());
+        String fwdKey = "fwd" + key;
+
         try {
-            getCache().set(key, 3600, ad);
+            /****** Add one single ad to fwd index ******/
+            getCache().set(fwdKey, 3600, ad);
```

Comment:
After some test, I got the results:
1) If we treating ad as an object in memcached such as getCache().set(fwdKey, 3600, ad); the memcached works like a HashMap which can be trusted. For instance, after running getCache().set(fwdKey, 3600, ad) twice, we still get only one key value pair <fwdKey, ad>.

2) It has to been checked if ad is added to an Object which is a container.  For example, we have key value pair <String key, Set<Ad> ads> and a single Ad newAd.  If we keep doing ads.add(newAd), we will have ads == {newAd, newAd} -> ads == {newAd, newAd, newAd} ->......  The newAd is treated as different objects by memcached. Note that Set cannot avoid the duplication.